### ISSUE TITLE
Is their a way to create custom tensor-like types in Keras (similar to torch.Tensor or tf.experimental.ExtensionType)

### ISSUE THREAD

**Issue by AyoubMDL** [keras-team/keras#21574]
Hi !
I’m wondering if there is currently any way to create a custom tensor type in Keras, similar to:

- `torch.Tensor` in PyTorch (subclassing Tensor)
- `tf.experimental.ExtensionType` in TensorFlow

The goal would be to create a framework-agnostic tensor type that works naturally with keras.ops and integrates into Keras layers without breaking backend compatibility.

Here is an example for illustration:
```python
import keras
import numpy as np


class QuantizedTensor(Tensor):
    def __init__(self, values: Tensor, scale: Tensor, value_bits: int):
        self.values = keras.ops.convert_to_tensor(values, dtype="float32")
        self.scale = keras.ops.convert_to_tensor(scale, dtype="float32")
        self.value_bits = value_bits


class QuantizedDense(keras.layers.Dense):
    def build(self, input_shape):
        return super().build(input_shape)

    def call(self, inputs):
        # Quantize the kernel
        self.kernel = self.quantizer(self.kernel)

        # MatMul between quantized tensors
        out = keras.ops.matmul(inputs.values, self.kernel.values)
        int_max = keras.ops.power(2.0, inputs.value_bits)
        out = keras.ops.clip(out, -int_max, int_max - 1)

        return QuantizedTensor(out, inputs.scale * self.kernel.scale)


values = np.ones((1, 32), dtype=np.float32)
scale = np.array(1.5, dtype=np.float32)
value_bits = 8
quantized_tensor = QuantizedTensor(values, scale, value_bits)

qlayer = QuantizedDense(units=10)
outputs = qlayer(quantized_tensor)
```

Thanks !

**Comment by dhantule** [comment]
Hi @AyoubMDL, thanks for reporting this. 

I've found a similar issue that discusses your problem please refer this [issue](https://github.com/keras-team/keras/issues/19230?utm_source=chatgpt.com) and let me know if it helps.

**Comment by AyoubMDL** [comment]
Hi @dhantule , thanks for your reply and for sharing the link. The issue you referenced mainly discusses type checking, whereas what I was specifically looking for was mentioned in this [comment](https://github.com/keras-team/keras/issues/19230#issuecomment-1971688428), particularly the idea of “a custom type that includes everything I want to be recognized as a tensor.”
As a workaround for the example I shared, I patched `keras.src.layers.layer.is_backend_tensor_or_symbolic` so that `QuantizedTensor` is treated as a tensor and can be passed to Keras layers. However, I’m not sure if this is the cleanest solution. Ideally, I’d like to declare something directly in the QuantizedTensor class so that it’s naturally recognized as a tensor.

Here’s the code with the patch applied:
```python
import keras
import numpy as np


class QuantizedTensor:
    def __init__(self, values, scale, value_bits):
        self.values = keras.ops.convert_to_tensor(values, dtype="float32")
        self.scale = keras.ops.convert_to_tensor(scale, dtype="float32")
        self.value_bits = value_bits

        self.dtype = keras.ops.dtype(self.values)
        self.shape = keras.ops.shape(self.values)


class QuantizedDense(keras.layers.Dense):
    def build(self, input_shape):
        return super().build(input_shape)

    def call(self, inputs):
        # Quantize the kernel
        # self.kernel = quantizer(self.kernel)

        # MatMul between quantized tensors
        out = keras.ops.matmul(inputs.values, self.kernel)
        int_max = keras.ops.power(2.0, inputs.value_bits)
        out = keras.ops.clip(out, -int_max, int_max - 1)

        return QuantizedTensor(out, inputs.scale, inputs.value_bits)


# Patch
original_is_tensor = keras.src.layers.layer.is_backend_tensor_or_symbolic


def custom_is_tensor(x, *args, **kwargs):
    if isinstance(x, QuantizedTensor):
        return True
    return original_is_tensor(x, *args, **kwargs)


keras.src.layers.layer.is_backend_tensor_or_symbolic = custom_is_tensor

# Create a quantized tensor
values = np.ones((1, 32), dtype=np.float32)
scale = np.array(1.5, dtype=np.float32)
value_bits = 8
quantized_tensor = QuantizedTensor(values, scale, value_bits)

# Forward on a keras layer
qlayer = QuantizedDense(units=10)
outputs = qlayer(quantized_tensor)
assert isinstance(outputs, QuantizedTensor)
```

**Comment by maitry63** [comment]
Hi @AyoubMDL,

That is a very clever workaround! Monkey-patching `keras.src.layers.layer.is_backend_tensor_or_symbolic` successfully tricks the layer input validator into letting your custom `QuantizedTensor` pass through.

However, relying on this in production comes with significant risks:
- Anything inside `keras.src` is internal. A minor update or patch release from the Keras team can rename or refactor this function, instantly breaking your code.
- While this bypasses the layer input check, other Keras utilities—like `model.save()`, multi-backend compilation, or gradient tracking will still fail or raise errors when they encounter an unrecognized custom class inside the execution graph.

As mentioned, Keras 3's backend-agnostic design means `keras.ops` and layers rely directly on backend-native tensors (`tf.Tensor`, `torch.Tensor`, `jax.Array`). Because JAX, PyTorch, and TensorFlow handle custom types completely differently (PyTrees vs. `torch.Tensor` subclasses vs. `ExtensionType`), Keras does not have a public, backend-agnostic dispatch mechanism for custom tensors.

If you want to avoid brittle internal patches, the standard approach remains using a plain Python container/dataclass and explicitly accessing `inputs.values`. Alternatively, if you are locked to a single backend, you can leverage native features , which let you pass custom objects naturally without touching Keras internals.

**Comment by AyoubMDL** [comment]
I see. Thanks @maitry63 for the answer. I'll close this issue.
